atc/db/pipeline_lifecycle.go

```diff
@@ -0,0 +1,89 @@
+package db
+
+import (
+	sq "github.com/Masterminds/squirrel"
+	"github.com/concourse/concourse/atc/db/lock"
+)
+
+//go:generate counterfeiter . PipelineLifecycle
+
+type PipelineLifecycle interface {
+	ArchiveAbandonedPipelines() (int, error)
+}
+
+func NewPipelineLifecycle(conn Conn, lockFactory lock.LockFactory) PipelineLifecycle {
+	return &pipelineLifecycle{
+		conn:        conn,
+		lockFactory: lockFactory,
+	}
+}
+
+type pipelineLifecycle struct {
+	conn        Conn
+	lockFactory lock.LockFactory
+}
+
+func (pl *pipelineLifecycle) ArchiveAbandonedPipelines() (int, error) {
+	tx, err := pl.conn.Begin()
+	if err != nil {
+		return 0, err
+	}
+
+	defer Rollback(tx)
+
+	latestSuccessfulBuilds := `WITH successful_builds AS (
+      SELECT DISTINCT ON (job_id) id, job_id
+      FROM builds
+      WHERE status='succeeded'
+      ORDER BY job_id, id DESC
+      )`
+	rows, err := pipelinesQuery.
+		Prefix(latestSuccessfulBuilds).
+		LeftJoin("jobs j ON j.id = p.parent_job_id").
+		LeftJoin("successful_builds ON successful_builds.job_id = p.parent_job_id").
+		Where(sq.Or{
```

Comment:
Not seeing a comment for this yet so dropping it here before I forget. We should also archive child pipelines whose parent pipeline has been archived. This was asked by Aidan on slack